### PR TITLE
s3 rusoto and general futures 0.3 migration of blobstore

### PR DESCRIPTION
**Pull request by drahnr**
https://github.com/mozilla/sccache/pull/869

### CODE REVIEW

**Review by TriplEight — DISMISSED** [review]
This can be merged as it's green in https://github.com/mozilla/sccache/pull/869

**Comment by drahnr** [comment]
Are we using s3 storage for our setup?

**Comment by TriplEight** [comment]
Nope, that would be quite slow and expensive. 
Though, there may be an option to organize our own s3 storage near the runners.

**Comment by drahnr** [comment]
Hence I am asking, minio scales pretty well.

File: src/cache/s3.rs [hunk 1]
```diff
@@ -13,52 +13,135 @@
...
-        let profile_providers = vec![
-            ProfileProvider::with_configuration(home.join(".aws").join("credentials"), "default"),
-            //TODO: this is hacky, this is where our mac builders store their
-            // credentials. We should either match what boto does more directly
-            // or make those builders put their credentials in ~/.aws/credentials
-            ProfileProvider::with_configuration(home.join(".boto"), "Credentials"),
-        ];
+        let profile_provider =
+            ProfileProvider::with_configuration(home.join(".aws").join("credentials"), "default")
+            // //TODO: this is hacky, this is where our mac builders store their
```
Comment (gww-parity) [inline src/cache/s3.rs:64]: double checking : are those double '//' '//' introduced intentionally?

File: src/dist/client_auth.rs [hunk 2]
```diff
@@ -627,15 +840,19 @@ pub fn get_token_oauth2_implicit(client_id: &str, mut auth_url: Url) -> Result<S
...
-        )
-    });
+    let shutdown_signal = shutdown_rx;
 
     let mut runtime = Runtime::new()?;
-    runtime.block_on(server.with_graceful_shutdown(shutdown_signal))?;
+    runtime.block_on(server.with_graceful_shutdown(async move {
+        let _ = shutdown_signal;
+    }))
+    // .map_err(|e| {
```
Comment (gww-parity) [inline src/dist/client_auth.rs:849]: do you want to leave this code commented?

is it ok to not handle this in new code?

File: src/bin/sccache-dist/token_check.rs [hunk 3]
```diff
@@ -152,22 +152,19 @@ impl MozillaCheck {
...
-            .context("Failed to interpret response from mozilla userinfo as string")?;
-        if !res.status().is_success() {
-            bail!(
-                "JWT forwarded to {} returned {}: {}",
-                url,
-                res.status(),
-                res_text
-            )
+            .context("Failed to interpret response from mozilla userinfo as string".to_owned())?;
+        if status.is_success() {
```
[the remark this answers is not in the document]
Comment (drahnr) [reply src/bin/sccache-dist/token_check.rs:160]: Good catch.

File: src/cache/s3.rs (hunk 1 above)
Comment (drahnr) [reply src/cache/s3.rs:64]: No, mit at all.

File: src/dist/client_auth.rs (hunk 2 above)
Comment (drahnr) [reply src/dist/client_auth.rs:849]: Its a bit of a TODO..

File: src/bin/sccache-dist/token_check.rs (hunk 3 above)
Comment (drahnr) [inline src/bin/sccache-dist/token_check.rs:160]:
```suggestion
        if !status.is_success() {
```

**Review by gww-parity — APPROVED** [review]

**Comment by drahnr** [comment]
Reducing the API surface was a very good call (thanks!).

Not sure how to proceed. rusoto has (I think) a few nice to have features.

I am bit wary of migrating to anything _super_-alpha mostly due to the potential breakage up the road, rusoto is likely going to be stable and stick around for quite some time.

It's your call essentially, not sure if other PRs like clean bucket depend on it.

Slight tendency to for rusoto, to simplify away some internal http chore we don't really care about.

**Comment by alnr** [comment]
Bump. Can this be merged?

**Comment by drahnr** [comment]
> Bump. Can this be merged?

This PR needs to be condensed down to the rusoto only chabgest, the features migration already landed in master. This is not a super high priority right now, but is also not that much work either so might be able to squeeze that in at some point in August.

**Comment by Xanewok** [comment]
@drahnr can you take a look at https://github.com/paritytech/cachepot/tree/igor-s3-rusoto? I adapted the changes from https://github.com/diem/sccache/tree/rusoto which AIUI the originally mentioned https://github.com/mozilla/sccache/pull/869 uses. I didn't force update the branch in case there's something else we might need to port as well.

**Pull request closed without merging**